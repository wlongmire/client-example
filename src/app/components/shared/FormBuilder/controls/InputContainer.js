import React from 'react'
import DynamicNumber from 'react-dynamic-number'
import isDefined from '../utils/isDefined'
import {
  HelpBlock,
  FormGroup,
  FormControl,
  ControlLabel,
  InputGroup,
  Tooltip,
  OverlayTrigger
} from 'react-bootstrap'

class InputContainer extends React.Component {
  constructor(props) {
    super(props)
    const name = this.props.data.name;
    this.state = {
      value: isDefined(this.props.initialValues[name]) ? this.props.initialValues[name] : '',
      isValid: null
    }

    this.handleChange = this.handleChange.bind(this)
    this.getValidationState = this.getValidationState.bind(this)
  }

  getValidationState() {

    // this bit of code acts as a circuit breaker
    // We want get validation to trigger a validation function and when results are back we need to call it again
    // But this breaks the loops
    if (this.state.isValid) {
      let isValid = this.state.isValid
      this.state.isValid = null
      return isValid
    }

    //only trigger validation if the value changes
    if (this.state.value !== '' &&
        this.props.data.attributes &&
        this.props.data.attributes.validationFunc) {
        this.props.validation[this.props.data.attributes.validationFunc](this.state.value).then((result)=> {
          this.setState({
            isValid: (result) ? 'success' : 'error'
          })
        })
    }

    if (this.state.isValid !== null) {
      return this.state.isValid
    }

    if (this.props.data.attributes && this.props.data.attributes.validationRegEx) {
      let regex = new RegExp(unescape(this.props.data.attributes.validationRegEx))
      return (regex.test(this.state.value)) ? 'success' : 'error'
    }

    return

  }

  handleChange(event) {
    this.setState({
      value: event.target.value
    });
  }

  render() {
    const tooltip = <Tooltip id={`tooltip_${this.props.data.questionId}`}> {this.props.data.tooltiptext}</Tooltip>
    let inputFormat = this.props.data.inputFormat
    if (inputFormat === 'currency') inputFormat = 'number'
    let input

    // Using a different input type for number and currency. It will format the number but the value will remain the raw number value
    if (['currency', 'number'].indexOf(this.props.data.inputFormat) > -1) {
      input = <DynamicNumber id={this.props.data.name} className="form-control number-control" separator={'.'} thousand={true} integer={1000} fraction={1000} onChange={this.handleChange} value={parseInt(this.state.value)}/>
    } else {
      input = (
        <FormControl
            placeholder={this.props.data.placeholder}
            type={inputFormat}
            onChange={this.handleChange}
            componentClass={(this.props.data.inputType === 'freeform') ? 'textarea' : 'input'}
            value={this.state.value}
          />
      )
    }

    const overlay = (
      <OverlayTrigger placement='top' overlay={tooltip} trigger={(this.props.data.tooltiptext) ? ['hover', 'focus'] : null}>
        {input}
      </OverlayTrigger>
    )

    const wrapper = (this.props.data.inputFormat === 'currency') ? <InputGroup><InputGroup.Addon>$</InputGroup.Addon>{overlay}</InputGroup> : overlay
    const helpBlock = (this.state.isValid === 'error') ? <HelpBlock>{this.props.data.validationMessage}</HelpBlock> : null;
    return(

       <FormGroup validationState={this.getValidationState()} controlId={this.props.data.name}>
        <ControlLabel>{this.props.data.text}</ControlLabel>
        {wrapper}
        {helpBlock}

       </FormGroup>
    )
  }
}

export default InputContainer