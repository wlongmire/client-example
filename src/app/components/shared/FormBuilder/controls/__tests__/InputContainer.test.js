/* eslint-env node, jest */

import React from 'react'
import { shallow } from 'enzyme'
import toJson from 'enzyme-to-json'
import { InputContainer } from '../InputContainer'

const dataProps = {
  attributes: {
    controlGroup: 'projectAddress'
  },
  inputType: 'text',
  inputFormat: 'text',
  name: 'projectCity',
  questionId: '3b',
  placeholder: 'City',
  required: true,
  tooltiptext: 'Please provide as descriptive of a street address as possible.'
}

describe('>>> Input Container Component --- Snapshot', () => {
  it('+++ provide a input field', () => {
    const renderedValue = shallow(
      <InputContainer
        user={{ broker: 'test-2222-7fd-b3ff-4fd3-9fc2-e752b9f5b002' }}
        initialValues={{}}
        initialParams={{}}
        data={dataProps}
      />)
    expect(toJson(renderedValue)).toMatchSnapshot()
  })
})


const initalParamsProps = {
  primaryInsuredAddress: { disabled: true },
  primaryInsuredCity: { disabled: true },
  primaryInsuredName: { disabled: true },
  primaryInsuredState: { disabled: true },
  primaryInsuredZipcode: { disabled: true },
  projectAddress: { disabled: true },
  projectCity: { disabled: true },
  projectState: { disabled: true }
}

const initalValuesProps = {
  clearanceStatus: 'pass',
  primaryInsuredAddress: '234234 23423489',
  primaryInsuredCity: '234234 23423489',
  primaryInsuredName: '234234 23423489',
  primaryInsuredState: 'District of Columbia',
  primaryInsuredZipcode: '234234 23423489',
  projectAddress: '234234 23423489',
  projectCity: '234234 23423489',
  projectState: 'California',
  projectZipcode: '234234 23423489',
  status: 'SUBMISSION',
  type: 'oi'
}

describe('>>> Input Container Component (with inital values) --- Snapshot', () => {
  it('+++ provide a input field (with inital values', () => {
    const renderedValue = shallow(
      <InputContainer
        user={{ broker: 'test-2222-7fd-b3ff-4fd3-9fc2-e752b9f5b002' }}
        initialValues={initalValuesProps}
        initialParams={initalParamsProps}
        data={dataProps}
      />)
    expect(toJson(renderedValue)).toMatchSnapshot()
  })
})