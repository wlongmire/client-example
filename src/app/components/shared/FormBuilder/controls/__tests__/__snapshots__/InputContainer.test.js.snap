// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`>>> Input Container Component (with inital values) --- Snapshot +++ provide a input field (with inital values 1`] = `
<FormGroup
  bsClass="form-group"
  controlId="projectCity"
>
  <OverlayTrigger
    defaultOverlayShown={false}
    overlay={
      <Tooltip
        bsClass="tooltip"
        className="in"
        id="tooltip_3b"
        placement="bottom"
      >
        Please provide as descriptive of a street address as possible.
      </Tooltip>
    }
    placement="top"
    trigger={
      Array [
        "hover",
        "focus",
      ]
    }
  >
    <FormControl
      autoComplete={false}
      bsClass="form-control"
      className="filled disabled"
      componentClass="input"
      disabled={true}
      id="projectCity"
      onChange={[Function]}
      placeholder="City"
      type="text"
      value="234234 23423489"
    />
  </OverlayTrigger>
  <AdditionalInfoComponent />
</FormGroup>
`;

exports[`>>> Input Container Component --- Snapshot +++ provide a input field 1`] = `
<FormGroup
  bsClass="form-group"
  controlId="projectCity"
>
  <OverlayTrigger
    defaultOverlayShown={false}
    overlay={
      <Tooltip
        bsClass="tooltip"
        className="in"
        id="tooltip_3b"
        placement="bottom"
      >
        Please provide as descriptive of a street address as possible.
      </Tooltip>
    }
    placement="top"
    trigger={
      Array [
        "hover",
        "focus",
      ]
    }
  >
    <FormControl
      autoComplete={false}
      bsClass="form-control"
      className=""
      componentClass="input"
      disabled={false}
      id="projectCity"
      onChange={[Function]}
      placeholder="City"
      type="text"
      value=""
    />
  </OverlayTrigger>
  <AdditionalInfoComponent />
</FormGroup>
`;
